### ISSUE TITLE
Nightly builds

### ISSUE THREAD

**Issue by omry** [facebookresearch/hydra#1112]
Automated job publishing a nightly package.
A few questions to answer:
- If can publish to pip:  
  - Any prior art?
  - Can we keep last X packages in pip without polluting the package version namespace?
- If not, http? can pip install have an http dependency?
- How to version the nightly packages?

**Comment by jieru-hu** [comment]
I looked into pytorch and ray a while ago on their nightly builds
Looks like neither of them publish to pip, they either host the whls on the website or S3

pytorch https://pytorch.org/
To install nightly builds
Builds are versioned by dates.
```
pip install numpy
pip install --pre torch torchvision torchaudio -f https://download.pytorch.org/whl/nightly/cu110/torch_nightly.html
```

ray
builds are versioned by commit SHAs and hosted on S3.
https://docs.ray.io/en/master/installation.html#installing-from-a-specific-commit
the latest nightly job will be hosted in the same location, "tagged" as latest, something like the following.
https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-1.0.0-cp37-cp37m-macosx_10_13_intel.whl

**Comment by omry** [comment]
I would like to understand why they don't publish to pip.
is this because they are too large and pip does not want to host them so many times?

**Comment by shagunsodhani** [comment]
> I would like to understand why they don't publish to pip.
> is this because they are too large and pip does not want to host them so many times?

That is the reason for PyTorch atleast. The packages are already much above the permissible sizes. For more details: https://github.com/pypa/packaging-problems/issues/96

A [quick search on pypi](https://pypi.org/search/?q=nightly) indicates that in general, nightly versions can be released on 
pypi.

Another option we could consider is maintaining the nightly builds on github itself.

I didnt understand the part about "can pip install have an http dependency". Did you mean "can the user download the packages from some http server?" Yes they can.

**Comment by omry** [comment]
for example, can you have a dependency on an http url in your requirements file or your setup.py?

**Comment by shagunsodhani** [comment]
yes we can have with `-f` flag. [Example](https://github.com/fairinternal/codistillation/blob/master/requirements.txt#L1)

**Comment by omry** [comment]
tensorflow has a different package name for nightly builds:
https://pypi.org/project/tf-nightly/

**Comment by shagunsodhani** [comment]
Hmm interesting. Each build is > 200 mb . Maybe we too can host it on Pypi

**Comment by omry** [comment]
I bet Google got some special deal with them, if you look carefully you will see that each version is multiple files and about 1GB.
I think the advantage of putting it in a different package is that it does not contaminate the package public versions namespace.

**Comment by omry** [comment]
I started a discussion at pypa/warehouse#8792 to see if pypa can eventually "official" support for nightly builds.

**Comment by rsokl** [comment]
Having nightly builds of Hydra (and omegaconf) would be great. Presently, hydra-zen tests against pre-releases, but there are new features (like support for `_partial_` in the instantiate API) that we would like to be testing against ASAP.

**Comment by Jasha10** [comment]
Thanks for bumping this @rsokl. I'll look into setting up build automation.

**Comment by omry** [comment]
Closing as not planned. The original need for downstream projects to test Hydra changes before a stable release is now served by intentional `.devN` releases on PyPI. Hydra's current release tooling builds, validates, smoke-installs, and publishes coordinated development releases through the reviewed Trusted Publishing workflow. Scheduling an immutable publication for every nightly build would add package-version churn and bypass that deliberate approval model without enough demonstrated benefit.
